text/0000-declarative-shadow-dom-polyfill.md

```diff
@@ -0,0 +1,346 @@
+
+### Invariants
+
+1. We would only support native shadow DOM when rendering declarative shadow DOM, synthetic shadow DOM will be out of the picture.
+2. The `getInnerHTML` method from the spec will not be polyfilled.
+3. The new `ElementInternals.shadowRoot` will not be polyfilled.
+4. For completeness, each polyfill removes the `<template>` element (and the inline `<script>` node when needed), so that the resulting tree is identical to the declarative output.
+
+### Implementations
+
+When comparing the proposed polyfill to the native implementation in Chrome, the main comparison points are:
+
+1. Performance: how fast does the polyfill transform LWC SSR rendered HTML into the proper DOM structure? There’s an extensive [analysis](https://github.com/mfreed7/declarative-shadow-dom/blob/master/README.md#performance) in the spec, comparing chrome native vs polyfills in chrome. For completion, the [Performance comparison in Chrome, Safari, and Firefox](#performance-comparison-in-chrome-safari-and-firefox) sections compare them in Safari and Firefox.
+2. Does the polyfill take advantage of the browser streaming capability? For example, when you load a large page like https://html.spec.whatwg.org/ on a slow connection, you don’t need to wait for the page to load entirely to see content, the content will show immediately after it is parsed. Some examples are provided in [this repository](https://github.com/jodarove/dec-shadow-dom-polyfill-spike#examples).
+3. Cumulative Layout Shift (CLS). CLS is the unexpected shifting of webpage elements while the page is still downloading, eg: Flash of Unstyled Content (FOUC). Compare the examples from [this repository](https://github.com/jodarove/dec-shadow-dom-polyfill-spike#examples).
+
+#### Single loop script shadow root attachment ([reference](https://github.com/mfreed7/declarative-shadow-dom/blob/master/README.md#baseline-2---single-script-based-shadow-root-attachment))
+
+Uses a single script at the end of the HTML that loops over all templates to attach the shadow roots and move the template contents into each root.
+
+This approach is optimized for speed - the entire page is parsed first, and then one script does all of the shadow root attachment and population. The advantage here is that the parser is not blocked for `script` on each shadow root. The major **disadvantages** are:
+
+1. Streaming is no longer possible. The entire page will consist of inert `<template>` elements that do not render, until the final script loops through and "converts" them into shadow roots.
+2. Content shifts will likely occur. The initially-parsed result, if rendered, will not contain the shadow content. Once the final loop runs, and shadow roots are attached, the page layout will likely change, leading to CLS problems.
```

Comment:
It's worth noting that all these downsides only exist for non-Chromium browsers. In Chromium, we will still get streaming and no content shifts.